### PR TITLE
Loopback plugin

### LINKED ISSUE #97: Loopback devices - UP by default on host-bridge?
Hey All,

On the host-bridge plugin the loopback device is down by default. Is there any particular reason for that or could we set it to up as part of the plugin itself? Maybe we introduce a conf value to handle whether or not to up the loopback device in the namespace?

### PR DESCRIPTION
**Pull request by zachgersh**
WRT #97

- Introduces ginkgo / gomega style testing (you don't need to look at the diff for that, just dependencies).
- Noop on delete (because you can just nuke a namespace with its `lo`
- Depending on if you merge the better error handling PR first - I may need to tweak to remove the godeps bit
- priv-net-run now doesn't up a loopback!

### CODE REVIEW

**Comment by zachgersh** [comment]
So - this is red due to the fact that you need to be able to mkdir on /var/run/netns and you get permission denied :/

**Comment by zachgersh** [comment]
I don't know how to make this go green on travis. Works fine inside of a docker image locally.

**Comment by steveeJ** [comment]
Thanks for this one! I think all you need to do is to `./test` to `sudo ./test` in the travis.yml. Please see if that helps and I'll finish reviewing this.

**Comment by zachgersh** [comment]
@steveeJ - thanks for the reminder that you could do that, been awhile since I have used travis!

Now, because we installed go without sudo you can't find it when sudo (hence the build fails again).

**Comment by zachgersh** [comment]
@steveeJ - good to go now!

**Comment by steveeJ** [comment]
Thanks @zachgersh! The new dependencies look like a rather big addition, what exactly do we gain from these?

**Comment by zachgersh** [comment]
Hey @steveeJ,

Good question!

Our team (CloudFoundry Container Networking) uses [Ginkgo](https://onsi.github.io/ginkgo/) and [Gomega](https://onsi.github.io/gomega/) to support BDD-style tests (e.g. our [vxlan plugin](https://github.com/cloudfoundry-incubator/ducati-cni-plugins/blob/a8387e57a765a27aea9c93aa7eaef3ea49b49399/cmd/vxlan/vxlan.go) and our [runc / cni adapter](https://github.com/cloudfoundry-incubator/guardian-cni-adapter/blob/21f9a6c9b0549216d528754f274b6e6bb70284f7/controller/controller.go#L75-L84).  We like this style because test descriptions are in plain English, which helps us do test-driven development.  The Ginkgo test runner is very mature (e.g. focusing, parallel testing, randomly permuting test order) and there are many Gomega matchers that simplify testing of common scenarios (e.g. `MatchJSON`).

Our [other PR](https://github.com/appc/cni/pull/120) adds test coverage elsewhere, using the same approach.

For what it’s worth, [Kubernetes uses Ginkgo and Gomega](https://github.com/kubernetes/kubernetes/blob/918ddbe/test/e2e/metrics_grabber_test.go#L27-L28) also.

Our team is actively using CNI and we expect to add more test coverage over time.  Using Ginkgo and Gomega in CNI would make that work much easier for us.

**Comment by steveeJ** [comment]
Thank you @zachgersh for the explanation. I think we can leverage a full-featured testing framework in CNI. Personally I had not used anything else than Go's built-in testing package but ginkgo looks like a good choice.

Can I ask you to squash the commits before I do a full code review?

**Comment by zachgersh** [comment]
@steveeJ look at for a squash shortly.  I am going to squash everything except the addition of the ginkgo / gomega dependencies. Should be easier for you to do a code review that way.

**Comment by zachgersh** [comment]
@steveeJ believe this should be good to go after the tests pass.

**Comment by zachgersh** [comment]
@steveeJ anything else required here?

**Comment by steveeJ** [comment]
ping @zachgersh, this needs a rebase now ;)

**Comment by zachgersh** [comment]
@steveeJ - sorry for the delay, rebased. Could we try to get this merged today? I am available to help and could jump on IRC if needed.

**Comment by steveeJ** [comment]
I think this is good to go. I changed the "Fixes" in the PR comment. Strictly seen it doesn't fix the mentioned issue, which is about upping the loopback interface *by default*, while PR allows to do this with a designated plugin.
